include/boost/url/impl/pct_encoding.ipp

```diff
@@ -133,6 +132,89 @@ pct_decode_unchecked(
     return dest - dest0;
 }
 
+BOOST_URL_DECL
+std::size_t
+validate_pct_encoding(
+    string_view s,
+    error_code& ec,
+    pct_decode_opts const& opt) noexcept
+{
+    const auto is_safe = [opt](char c) {
+        return c != '%' &&
+               (opt.allow_null ||
+                c != '\0');
+    };
+
+    std::size_t pcts = 0;
+    char const* it = s.data();
+    char const* end = it + s.size();
+    it = grammar::find_if_not(it, end, is_safe);
```

Comment:
`allow_null` can't change during the find loop but you are checking it every time